File: backend/app/schemas/properties.py
"""
app/schemas/properties.py

Pydantic read shapes for the public property listing resources.
"""

from datetime import datetime
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, computed_field

from app.models.enums import (
    Furnishing,
    ListingType,
    MediaType,
    PaymentStructure,
    PriceFlexibility,
    PropertyStatus,
    PropertyType,
    VerificationStatus,
)
from app.schemas._pagination import PaginatedResponse


class PropertyMediaRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    media_type: MediaType
    url: str
    stream_uid: Optional[str] = None
    position: int
    is_cover: bool
    width: Optional[int] = None
    height: Optional[int] = None


class PlotDetails(BaseModel):
    """Sub-fields shown only when property_type is "plot" — stored in Property.plot_details (JSONB)."""

    model_config = ConfigDict(from_attributes=True)

    dimension: Optional[str] = Field(default=None, max_length=50)  # e.g. "30x40"
    is_corner_plot: Optional[bool] = None


class LandDetails(BaseModel):
    """Sub-fields shown only when property_type is "land" — stored in Property.land_details (JSONB)."""

    model_config = ConfigDict(from_attributes=True)

    land_use: Optional[str] = None  # "residential" | "commercial"
    approvals: list[str] = []  # e.g. ["RERA", "BMRDA"]


class JVPartner(BaseModel):
    """One row of Property.jv_details.partners — a joint-venture co-owner and their stake."""

    model_config = ConfigDict(from_attributes=True)

    name: str = Field(min_length=1, max_length=150)
    role: Optional[str] = Field(default=None, max_length=100)
    split_percent: Optional[float] = Field(default=None, ge=0, le=100)
    email: Optional[str] = Field(default=None, max_length=255)
    can_edit: bool = False
    can_approve: bool = False
    can_publish: bool = False


class JVDetails(BaseModel):
    """
    Sub-fields shown only when Property.is_jv_property is true — stored in
    Property.jv_details (JSONB). Not tied to a specific property_type: the
    Figma design shows the same "Is this a JV Property?" toggle on every
    Sell property type's Step 1 form.
    """

    model_config = ConfigDict(from_attributes=True)

    partners: list[JVPartner] = []
    commission_mode: Optional[Literal["auto", "manual"]] = None
    # Object key of the uploaded JV agreement document (private bucket, not
    # a public URL) — set by POST /properties/{id}/jv-agreement, after the
    # property already exists.
    agreement_document_key: Optional[str] = None


class PGDetails(BaseModel):
    """
    Sub-fields shown only when property_type is "pg_colive" — stored in
    Property.pg_details (JSONB). Covers three distinct Figma sub-forms in
    one flexible blob (same reasoning as PlotDetails/LandDetails: read and
    written as a whole, never filtered on):
      - Sell: building-level fields (listing_scope is None)
      - Rent, "Entire Building" (listing_scope="entire")
      - Rent, "Unit / Room" (listing_scope="unit")
    """

    model_config = ConfigDict(from_attributes=True)

    listing_scope: Optional[Literal["entire", "unit"]] = None

    # Sell: PG/co-living building details.
    total_floors: Optional[int] = None
    currently_operational: Optional[bool] = None
    estimated_monthly_revenue: Optional[float] = None

    # Shared across Sell and Rent/Entire.
    total_rooms: Optional[int] = None
    occupancy_types: list[str] = []  # e.g. ["Single", "Double", "Triple"]
    gender: Optional[str] = None  # "Male" | "Female" | "Mixed"

    # Rent/Entire only.
    monthly_rent_per_bed: Optional[float] = None

    # Rent/Unit only.
    room_type: Optional[str] = None
    floor: Optional[int] = None
    bathroom_type: Optional[str] = None  # "Attached" | "Common"
    ac: Optional[str] = None  # "AC" | "Non-AC"
    gender_preference: Optional[str] = None  # "Male" | "Female" | "Any"
    monthly_rent: Optional[float] = None

    # Shared across both Rent sub-forms.
    meals_included: Optional[bool] = None
    amenities: list[str] = []


class PropertyBrokerVerification(BaseModel):
    """Embedded on PropertyBrokerRead only when the broker has a profile."""

    model_config = ConfigDict(from_attributes=True)

    verification_status: VerificationStatus


class PropertyBrokerRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    full_name: Optional[str] = None
    broker_profile: Optional[PropertyBrokerVerification] = None


class PropertyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    description: Optional[str] = None
    status: PropertyStatus
    listing_type: ListingType
    property_type: PropertyType
    price: float
    price_per_sqft: Optional[float] = None
    token_amount: Optional[float] = None
    maintenance_monthly: Optional[float] = None
    deposit: Optional[float] = None
    is_negotiable: bool
    price_flexibility: Optional[PriceFlexibility] = None
    payment_structure: Optional[PaymentStructure] = None
    stamp_duty_percent: Optional[float] = None
    registration_fee_percent: Optional[float] = None
    brokerage_included: bool
    brokerage_percent: Optional[float] = None
    bhk: Optional[int] = None
    bathrooms: Optional[int] = None
    area_sqft: Optional[float] = None
    floor: Optional[int] = None
    total_floors: Optional[int] = None
    facing: Optional[str] = None
    built_year: Optional[int] = None
    parking_slots: Optional[int] = None
    furnishing: Optional[Furnishing] = None
    amenities: list[str]
    plot_details: Optional[PlotDetails] = None
    land_details: Optional[LandDetails] = None
    pg_details: Optional[PGDetails] = None
    is_jv_property: bool
    jv_details: Optional[JVDetails] = None
    virtual_tour_url: Optional[str] = None
    address_line: str
    locality: str
    city: str
    state: str
    pincode: str
    landmark: Optional[str] = None
    metro_distance_m: Optional[int] = None
    media: list[PropertyMediaRead]
    broker: PropertyBrokerRead
    published_at: Optional[datetime] = None

    @computed_field
    @property
    def metro_distance_km(self) -> Optional[float]:
        if self.metro_distance_m is None:
            return None
        return round(self.metro_distance_m / 1000, 1)


class PropertyListItem(BaseModel):
    """One card's worth of data for the /properties search results grid — lighter than PropertyRead, no media gallery or broker detail."""

    id: UUID
    title: str
    listing_type: ListingType
    property_type: PropertyType
    price: float
    bhk: Optional[int] = None
    bathrooms: Optional[int] = None
    area_sqft: Optional[float] = None
    furnishing: Optional[Furnishing] = None
    city: str
    locality: str
    cover_image_url: Optional[str] = None
    published_at: Optional[datetime] = None


class PropertyListResponse(PaginatedResponse[PropertyListItem]):
    """Pagination envelope for GET /properties: page/page_size in, total/total_pages computed for the caller."""


class BrokerPropertyListItem(PropertyListItem):
    """
    GET /properties/mine's card shape — PropertyListItem plus status, since
    a broker (unlike the public search grid) needs to see draft/pending
    listings too, not just active ones.
    """

    status: PropertyStatus


class PropertyCompareResponse(BaseModel):
    """GET /properties/compare — a normalized spec table, reusing PropertyRead's full field set."""

    items: list[PropertyRead]


class NeighborhoodSummary(BaseModel):
    """One locality's worth of data for the search overlay's Curated Neighborhoods grid."""

    locality: str
    city: str
    property_count: int
    cover_image_url: Optional[str] = None


class PropertyCreateRequest(BaseModel):
    """
    The Post Property wizard's full submission — Step 1 (Property Info) +
    Step 3 (Pricing) fields together in one request. Steps 1 and 2 only
    collect data client-side; nothing is persisted until this fires,
    because `Property.price` is NOT NULL with a `price > 0` check
    constraint, so a valid row can't exist before pricing is known.
    Supports residential listings (apartment/villa/independent_house),
    Plot, Land, PG/co-living (Sell and Rent, including Rent's Entire
    Building / Unit-Room split), and the Rent-only commercial types
    (shop/commercial_building/built_to_suit) — plus a cross-cutting JV
    toggle available on any Sell property type. The schema itself doesn't
    enforce which property_type values are "allowed" for a given
    listing_type; the wizard's frontend does, by only offering the
    combinations it has a Step 1 sub-form for.
    """

    title: str = Field(min_length=1, max_length=200)
    listing_type: ListingType
    property_type: PropertyType
    bhk: Optional[int] = None
    bathrooms: Optional[int] = None
    area_sqft: Optional[float] = Field(default=None, gt=0)
    facing: Optional[str] = Field(default=None, max_length=20)
    furnishing: Optional[Furnishing] = None
    built_year: Optional[int] = None
    amenities: list[str] = []
    plot_details: Optional[PlotDetails] = None
    land_details: Optional[LandDetails] = None
    pg_details: Optional[PGDetails] = None
    is_jv_property: bool = False
    jv_details: Optional[JVDetails] = None
    virtual_tour_url: Optional[str] = Field(default=None, max_length=500)
    address_line: str = Field(min_length=1, max_length=255)
    locality: str = Field(min_length=1, max_length=100)
    city: str = Field(min_length=1, max_length=100)
    state: str = Field(min_length=1, max_length=100)
    pincode: str = Field(min_length=1, max_length=6)
    landmark: Optional[str] = None
    price: float = Field(gt=0)
    price_per_sqft: Optional[float] = Field(default=None, gt=0)
    token_amount: Optional[float] = Field(default=None, gt=0)
    maintenance_monthly: Optional[float] = Field(default=None, gt=0)
    deposit: Optional[float] = Field(default=None, gt=0)
    is_negotiable: bool = False
    price_flexibility: Optional[PriceFlexibility] = None
    payment_structure: Optional[PaymentStructure] = None
    stamp_duty_percent: Optional[float] = Field(default=None, ge=0)
    registration_fee_percent: Optional[float] = Field(default=None, ge=0)
    brokerage_included: bool = True
    brokerage_percent: Optional[float] = Field(default=None, ge=0)

File: frontend/lib/api/endpoints/properties.ts
// lib/api/endpoints/properties.ts
// Typed functions for the public /properties resource. No auth required —
// GET /properties/{id} serves the public Property Details screen,
// GET /properties serves the Listings search grid.

import { apiRequest, apiRequestMultipart } from "@/lib/api/client";
import type { Furnishing, ListingType, MediaType, PropertyStatus, PropertyType, VerificationStatus } from "@/lib/enums";
import type { LandDetailsValues, PlotDetailsValues } from "@/lib/validation/postProperty";

export type PropertyMediaRead = {
  id: string;
  media_type: MediaType;
  url: string;
  stream_uid?: string;
  position: number;
  is_cover: boolean;
  width?: number;
  height?: number;
};

export type PropertyBrokerRead = {
  id: string;
  full_name?: string;
  broker_profile?: {
    verification_status: VerificationStatus;
  };
};

export type PropertyRead = {
  id: string;
  title: string;
  description?: string;
  status: PropertyStatus;
  listing_type: ListingType;
  property_type: PropertyType;
  price: number;
  maintenance_monthly?: number;
  deposit?: number;
  is_negotiable: boolean;
  bhk?: number;
  bathrooms?: number;
  area_sqft?: number;
  floor?: number;
  total_floors?: number;
  facing?: string;
  built_year?: number;
  parking_slots?: number;
  furnishing?: Furnishing;
  amenities: string[];
  plot_details?: PlotDetailsValues;
  land_details?: LandDetailsValues;
  address_line: string;
  locality: string;
  city: string;
  state: string;
  pincode: string;
  landmark?: string;
  metro_distance_m?: number;
  metro_distance_km?: number;
  media: PropertyMediaRead[];
  broker: PropertyBrokerRead;
  published_at?: string;
};

export function getProperty(id: string): Promise<PropertyRead> {
  return apiRequest<PropertyRead>(`/properties/${id}`);
}

export type PropertySort = "newest" | "price_asc" | "price_desc";

export type PropertyListItem = {
  id: string;
  title: string;
  listing_type: ListingType;
  property_type: PropertyType;
  price: number;
  // Pydantic's Optional[T] always serializes as an explicit JSON `null`,
  // never an absent key — typed `| null` (not `?`) so a null !== undefined
  // check downstream can't silently treat "no value" as "has a value".
  bhk: number | null;
  bathrooms: number | null;
  area_sqft: number | null;
  furnishing: Furnishing | null;
  city: string;
  locality: string;
  cover_image_url: string | null;
  published_at: string | null;
};

export type PropertyListResponse = {
  items: PropertyListItem[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
};

export type PropertyListParams = {
  city?: string;
  locality?: string;
  // Free-text match against title, description, city, locality, landmark,
  // or amenities (matches ANY of them) — for the nav search box, which
  // can't know what kind of value was typed. Distinct from city/locality
  // above, which are exact matches driven by dropdowns and neighborhood
  // links that already know the precise value. Deliberately just
  // substring matching, not keyword/facet parsing — property type/bedroom
  // count already have precise dedicated filters below.
  search?: string;
  listing_type?: ListingType;
  property_type?: PropertyType[];
  price_min?: number;
  price_max?: number;
  bhk_min?: number;
  amenities?: string[];
  sort?: PropertySort;
  page?: number;
  page_size?: number;
};

// The param shape buildQueryString actually serializes — a superset of
// PropertyListParams (`sort` widened to `string`) so other list endpoints
// with their own sort literal union (e.g. saved-properties' SavedSort) can
// reuse this same serialization instead of hand-rolling their own
// URLSearchParams construction. Every concrete params type below (a
// PropertyListParams, a SavedPropertyListParams, ...) is a structural
// subtype of this and can be passed directly.
export type ListQueryParams = {
  city?: string;
  locality?: string;
  search?: string;
  listing_type?: ListingType;
  property_type?: PropertyType[];
  price_min?: number;
  price_max?: number;
  bhk_min?: number;
  amenities?: string[];
  sort?: string;
  page?: number;
  page_size?: number;
};

// Exported so the Listings page can build the identical query string for
// the browser URL (shareable/bookmarkable filtered views), and so other
// /list-style endpoints (e.g. saved-properties) can reuse the same
// param-serialization logic instead of duplicating it.
export function buildQueryString(params: ListQueryParams): string {
  const queryParams = new URLSearchParams();
  if (params.city) queryParams.set("city", params.city);
  if (params.locality) queryParams.set("locality", params.locality);
  if (params.search) queryParams.set("search", params.search);
  if (params.listing_type) queryParams.set("listing_type", params.listing_type);
  for (const value of params.property_type ?? []) queryParams.append("property_type", value);
  if (params.price_min !== undefined) queryParams.set("price_min", String(params.price_min));
  if (params.price_max !== undefined) queryParams.set("price_max", String(params.price_max));
  if (params.bhk_min !== undefined) queryParams.set("bhk_min", String(params.bhk_min));
  for (const value of params.amenities ?? []) queryParams.append("amenities", value);
  if (params.sort) queryParams.set("sort", params.sort);
  if (params.page !== undefined) queryParams.set("page", String(params.page));
  if (params.page_size !== undefined) queryParams.set("page_size", String(params.page_size));
  const query = queryParams.toString();
  return query ? `?${query}` : "";
}

export function listProperties(params: PropertyListParams = {}): Promise<PropertyListResponse> {
  return apiRequest<PropertyListResponse>(`/properties${buildQueryString(params)}`);
}

export const MAX_COMPARE_IDS = 3;

export type PropertyCompareResponse = {
  items: PropertyRead[];
};

export function compareProperties(ids: string[]): Promise<PropertyCompareResponse> {
  return apiRequest<PropertyCompareResponse>(`/properties/compare?ids=${ids.join(",")}`);
}

export type NeighborhoodSummary = {
  locality: string;
  city: string;
  property_count: number;
  cover_image_url: string | null;
};

export function getNeighborhoods(limit = 4): Promise<NeighborhoodSummary[]> {
  return apiRequest<NeighborhoodSummary[]>(`/properties/neighborhoods?limit=${limit}`);
}

// Broker-authenticated Post Property wizard. createProperty fires once,
// at the wizard's final step — Steps 1 (info) and 2 (media) are only
// collected client-side until then, since the backend can't create a
// valid row before price is known (Property.price is NOT NULL with a
// `price > 0` check). See PropertyCreateRequest's docstring in
// app/schemas/properties.py for the full reasoning.
export type PropertyCreateInput = {
  title: string;
  listing_type: ListingType;
  property_type: PropertyType;
  bhk?: number;
  bathrooms?: number;
  area_sqft?: number;
  facing?: string;
  furnishing?: Furnishing;
  built_year?: number;
  amenities: string[];
  plot_details?: PlotDetailsValues;
  land_details?: LandDetailsValues;
  address_line: string;
  locality: string;
  city: string;
  state: string;
  pincode: string;
  landmark?: string;
  price: number;
  maintenance_monthly?: number;
  deposit?: number;
  is_negotiable: boolean;
};

export function createProperty(data: PropertyCreateInput): Promise<PropertyRead> {
  return apiRequest<PropertyRead>("/properties", { method: "POST", body: data });
}

export function uploadPropertyMedia(propertyId: string, images: File[]): Promise<PropertyMediaRead[]> {
  const formData = new FormData();
  for (const image of images) formData.append("images", image);
  return apiRequestMultipart<PropertyMediaRead[]>(`/properties/${propertyId}/media`, formData);
}

export function submitProperty(propertyId: string): Promise<PropertyRead> {
  return apiRequest<PropertyRead>(`/properties/${propertyId}/submit`, { method: "POST" });
}
